### ISSUE TITLE
WebSocket Streaming API improvement

### ISSUE THREAD

**Issue by akihikodaki** [tootsuite/mastodon#3049]
# Flaws of the Current WebSocket Streaming API API
The current WebSocket streaming API has the following flaws

## `payload` Encoding
`payload` is a JSON object encoded as a string for some events, but it is just a plain JSON object in the others.

For example,
```JSON
{"payload":"{\"id\":1,\"created_at\":\"2000-01-01T00:00:00.000Z\",\"in_reply_to_id\":null,\"in_reply_to_account_id\":null,\"sensitive\":false,\"spoiler_text\":\"\",\"visibility\":\"public\",\"application\":{\"name\":\"Web\",\"website\":null},\"account\":{\"id\":1,\"username\":\"admin\",\"acct\":\"admin\",\"display_name\":\"\",\"locked\":false,\"created_at\":\"2000-01-01T00:00:00.000Z\",\"followers_count\":0,\"following_count\":0,\"statuses_count\":0,\"note\":\"\",\"url\":\"https://cb6e6126.ngrok.io/@admin\",\"avatar\":\"https://cb6e6126.ngrok.io/headers/original/missing.png\",\"avatar_static\":\"https://cb6e6126.ngrok.io/headers/original/missing.png\",\"header\":\"https://cb6e6126.ngrok.io/headers/original/missing.png\",\"header_static\":\"https://cb6e6126.ngrok.io/headers/original/missing.png\"},\"media_attachments\":[],\"mentions\":[],\"tags\":[],\"uri\":\"tag:cb6e6126.ngrok.io,2000-01-01:objectId=1:objectType=Status\",\"content\":\"<p>hello, world</p>\",\"url\":\"https://cb6e6126.ngrok.io/@admin/1\",\"reblogs_count\":0,\"favourites_count\":0,\"reblog\":null,\"favourited\":false,\"reblogged\":false}","event":"update"}
```
```JSON
{"payload":1,"event":"delete"}
```

Pull request #2609 tried to resolve the problem, but it is abandoned because fixing this in the existing API will be breaking.

## Endpoints Inconsistency with the Event Stream API
The event stream API has endpoints such as `/api/v1/strereaming/user`, but the WebSocket counterpart has `/api/v1/streaming?stream=user`. That is inconsistent.

# Possible Improvement
I have an idea to create a new API with endpoints the event stream uses, which fixes `payload` as well.
In this case, `/api/v1/streaming/user` emits an event like this:
```JSON
{"payload":{"id":1,"created_at":"2000-01-01T00:00:00.000Z","in_reply_to_id":null,"in_reply_to_account_id":null,"sensitive":false,"spoiler_text":"","visibility":"public","application":{"name":"Web","website":null},"account":{"id":1,"username":"admin","acct":"admin","display_name":"","locked":false,"created_at":"2000-01-01T00:00:00.000Z","followers_count":0,"following_count":0,"statuses_count":0,"note":"","url":"https://cb6e6126.ngrok.io/@admin","avatar":"https://cb6e6126.ngrok.io/headers/original/missing.png","avatar_static":"https://cb6e6126.ngrok.io/headers/original/missing.png","header":"https://cb6e6126.ngrok.io/headers/original/missing.png","header_static":"https://cb6e6126.ngrok.io/headers/original/missing.png"},"media_attachments":[],"mentions":[],"tags":[],"uri":"tag:cb6e6126.ngrok.io,2000-01-01:objectId=1:objectType=Status","content":"<p>hello, world</p>","url":"https://cb6e6126.ngrok.io/@admin/1","reblogs_count":0,"favourites_count":0,"reblog":null,"favourited":false,"reblogged":false},"event":"update"}
```
While the older counterpart remains at `/api/v1/streaming?stream=user`:
```JSON
{"payload":"{\"id\":1,\"created_at\":\"2000-01-01T00:00:00.000Z\",\"in_reply_to_id\":null,\"in_reply_to_account_id\":null,\"sensitive\":false,\"spoiler_text\":\"\",\"visibility\":\"public\",\"application\":{\"name\":\"Web\",\"website\":null},\"account\":{\"id\":1,\"username\":\"admin\",\"acct\":\"admin\",\"display_name\":\"\",\"locked\":false,\"created_at\":\"2000-01-01T00:00:00.000Z\",\"followers_count\":0,\"following_count\":0,\"statuses_count\":0,\"note\":\"\",\"url\":\"https://cb6e6126.ngrok.io/@admin\",\"avatar\":\"https://cb6e6126.ngrok.io/headers/original/missing.png\",\"avatar_static\":\"https://cb6e6126.ngrok.io/headers/original/missing.png\",\"header\":\"https://cb6e6126.ngrok.io/headers/original/missing.png\",\"header_static\":\"https://cb6e6126.ngrok.io/headers/original/missing.png\"},\"media_attachments\":[],\"mentions\":[],\"tags\":[],\"uri\":\"tag:cb6e6126.ngrok.io,2000-01-01:objectId=1:objectType=Status\",\"content\":\"<p>hello, world</p>\",\"url\":\"https://cb6e6126.ngrok.io/@admin/1\",\"reblogs_count\":0,\"favourites_count\":0,\"reblog\":null,\"favourited\":false,\"reblogged\":false}","event":"update"}
```

# Discussion
If you have a solution different I have came up, please write a comment. Also, I would appreciate additional improvements we could have when we introduces a new API as I suggested.

* * * *

- [X] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [X] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by fermuch** [comment]
If the idea is to release a new backwards-incompatible API, why not go with [protobuf](https://github.com/google/protobuf/tree/master/ruby) instead of JSON? Faster parsing, less bytes to transfer, and usable in almost any language[¹](http://blog.codeclimate.com/blog/2014/06/05/choose-protocol-buffers/).

**Comment by akihikodaki** [comment]
For your information, the streaming API server is implemented in Node.js. Protocol buffers should be available anyway.

**Comment by beatrix-bitrot** [comment]
@akihikodaki i like it. are there any other things you would want to change?

also, should this still be /api/v1 ? or /api/v1.x or /api/v2? etc?

**Comment by akihikodaki** [comment]
In my opinion, `/api/v1` should be fine if this API change involves only the streaming API.
When a API change involves many other APIs, it may be appropriate to have `/api/v1+` and allow user agents to assume the new ones are available when one of them returned non-404 code. In this case, however, this change alters only the streaming API and it would not benefit much to change the version.

**Comment by akihikodaki** [comment]
After some thinking, I'm a bit against with protocol buffers.

> Faster parsing, less bytes to transfer, and usable in almost any language.

That would be true, but I have two counter arguments:
* Library is required to be downloaded
Though it may reduce transfers in the streaming, users may have to download the library. For example, there is [an argument that the JavaScript of Mastodon is already too big](https://github.com/tootsuite/mastodon/pull/2981#pullrequestreview-38003111).

* It could be still burden for developers
Or JSON is so easy to use for most developers. Many programming language provides JSON decoders by default. The advantage is hard to beat.

In short, protocol buffers is not _always_ good and we probably need some justification to adopt.

**Comment by akihikodaki** [comment]
I'm wondering we could add an option to include binary media if we adopt protocol buffers, though I'm not sure streaming server can deal with the burden.

**Comment by ThisIsMissEm** [comment]
You could also look into GraphQL + it's subscriptions infrastructure, which would allow clients to just request the fields they need. Though it'd be a lot more complex to implement at this stage.

**Comment by ThisIsMissEm** [comment]
If we can come to an agreed direction for this, then I'd be happy to do the work on it; I've been wanting to refactor the streaming server code to be more readable anyway.
